pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/AvgShedder.java

```diff
@@ -0,0 +1,302 @@
+                            + "minMsgThreshold:{}, minThroughputThreshold:{}",
+                    highThreshold, lowThreshold, hitCountHighThreshold, hitCountLowThreshold,
+                    minMsgThreshold, minThroughputThreshold);
+        }
+
+        List<String> brokers = new ArrayList<>(loadData.getBrokerData().size());
+        brokerScoreMap.clear();
+        // calculate scores of brokers.
+        for (Map.Entry<String, BrokerData> entry : loadData.getBrokerData().entrySet()) {
+            LocalBrokerData localBrokerData = entry.getValue().getLocalData();
+            String broker = entry.getKey();
+            brokers.add(broker);
+            Double score = calculateScores(localBrokerData, conf);
+            brokerScoreMap.put(broker, score);
+            if (log.isDebugEnabled()) {
+                log.info("broker:{}, scores:{}, throughput:{}, messageRate:{}", broker, score,
+                        localBrokerData.getMsgThroughputIn() + localBrokerData.getMsgThroughputOut(),
+                        localBrokerData.getMsgRateIn() + localBrokerData.getMsgRateOut());
+            }
+        }
+
+        // sort brokers by scores.
+        brokers.sort((e1, e2) -> (int) (brokerScoreMap.get(e1) - brokerScoreMap.get(e2)));
+        log.info("sorted broker list:{}", brokers);
+
+        // find broker pairs for shedding.
+        List<Pair<String, String>> pairs = new LinkedList<>();
+        int i = 0, j = brokers.size() - 1;
+        while (i <= j) {
+            String maxBroker = brokers.get(j);
+            String minBroker = brokers.get(i);
+            if (brokerScoreMap.get(maxBroker) - brokerScoreMap.get(minBroker) < lowThreshold) {
+                brokerHitCountForHigh.remove(maxBroker);
+                brokerHitCountForHigh.remove(minBroker);
+
+                brokerHitCountForLow.remove(maxBroker);
+                brokerHitCountForLow.remove(minBroker);
+            } else {
+                pairs.add(Pair.of(minBroker, maxBroker));
+                if (brokerScoreMap.get(maxBroker) - brokerScoreMap.get(minBroker) < highThreshold) {
+                    brokerHitCountForLow.put(minBroker, brokerHitCountForLow.getOrDefault(minBroker, 0) + 1);
+                    brokerHitCountForLow.put(maxBroker, brokerHitCountForLow.getOrDefault(maxBroker, 0) + 1);
+
+                    brokerHitCountForHigh.remove(maxBroker);
+                    brokerHitCountForHigh.remove(minBroker);
+                } else {
+                    brokerHitCountForLow.put(minBroker, brokerHitCountForLow.getOrDefault(minBroker, 0) + 1);
+                    brokerHitCountForLow.put(maxBroker, brokerHitCountForLow.getOrDefault(maxBroker, 0) + 1);
+
+                    brokerHitCountForHigh.put(minBroker, brokerHitCountForHigh.getOrDefault(minBroker, 0) + 1);
+                    brokerHitCountForHigh.put(maxBroker, brokerHitCountForHigh.getOrDefault(maxBroker, 0) + 1);
```

Comment:
It would be better to use `Map<String, MutableInt>` for `brokerHitCountForLow` and `brokerHitCountForHigh`. Then you can update the map like:

```java
map.computeIfAbsent(key, __ -> new MutableInt(0)).increment();
```

The current code uses 1 `get` operation and 1 `put` operation each time no matter if the key exists. The code above only uses 1 `get` operation for existing values or 1 `put` operation for non-existing values.

Reply:
Done, Thanks for review!